### PR TITLE
Pawel/signed url

### PR DESCRIPTION
**Pull request by pawelsz-rb**
## Description of the change
- Supporting uploads through signed URL, leaving old way of uploading intact. 
- The new way will zip all the map files and upload one zip file via signed URL received by our API
-  In addition, it will create a manifest.json file and store project_id, minified_url, and code version, which is needed by other processes down the road.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#79437]() 
   Fix [#79561]()

### Development

- [x] Lint rules pass locally
- [x] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review 

- [x]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [x] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [x] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/common/rollbar-api.js [hunk 1]
```diff
@@ -36,18 +35,33 @@ class RollbarAPI {
...
   }
 
+  async sigendURLsourcemaps(request) {
+
+    const json = JSON.stringify({ version: request.version , prefix_url: request.baseUrl});
+
+    const resp = await this.axios.post(
+      '/signed_url/sourcemaps', json,
+        {headers: {
+            // axios needs some help with headers for form data.
```
Comment (waltjones) [inline src/common/rollbar-api.js]: This comment is out of place, as the request here doesn't use form data or its headers.

File: src/common/rollbar-api.js [hunk 2]
```diff
@@ -36,18 +35,33 @@ class RollbarAPI {
     return this.processResponse(resp);
   }
 
+  async sigendURLsourcemaps(request) {
+
+    const json = JSON.stringify({ version: request.version , prefix_url: request.baseUrl});
```
Comment (waltjones) [inline src/common/rollbar-api.js:41]: Axios can accept the javascript object directly, and when doing so, will automatically set the content type. The `JSON.stringify` and `headers` can be removed, and these lines can be simplified to just:
```
const resp = await this.axios.post(
  '/signed_url/sourcemaps',
  { version: request.version , prefix_url: request.baseUrl}
);
```

File: src/common/rollbar-api.js [hunk 3]
```diff
@@ -36,18 +35,33 @@ class RollbarAPI {
...
-        'Content-Type': `multipart/form-data; boundary=${form.getBoundary()}`,
-        'Content-Length': form.getLengthSync()
-      }}
+        '/sourcemap',
+        form.getBuffer(), // use buffer to prevent unwanted string escaping.
+        {headers: {
+            // axios needs some help with headers for form data.
+            'Content-Type': `multipart/form-data; boundary=${form.getBoundary()}`,
+            'Content-Length': form.getLengthSync()
+          }}
```
Comment (waltjones) [inline src/common/rollbar-api.js]: This looks like incorrect indentation.

File: src/sourcemaps/command.js [hunk 4]
```diff
@@ -35,6 +37,12 @@ exports.builder = function (yargs) {
     type: 'boolean',
     demandOption: false
   })
+      .option('signed-url', {
+        describe: 'Zip all the source map files and upload via signed URL',
+        requiresArg: false,
+        type: 'boolean',
+        demandOption: false
+      })
```
Comment (waltjones) [inline src/sourcemaps/command.js]: Indentation also needs fixed here.

File: src/sourcemaps/command.js [hunk 5]
```diff
@@ -50,13 +58,32 @@ exports.handler = async function (argv) {
...
+    })
+
+    const signedUrlUploader = new SignedUrlUploader()
+    await requester.requestSignedUrl(argv['dry-run'])
+    if (requester.data && requester.data['err'] === 0) {
+      requester.setProjectID(requester.data['result']['project_id'])
+      const manifestFile = requester.createManifestFile(scanner.targetPath)
+      signedUrlUploader.mapFiles(scanner.files);
+      signedUrlUploader.zipFiles(scanner.targetPath, requester.data['result']['filename'], manifestFile);
+      await signedUrlUploader.upload(argv['dry-run'], requester.data['result']['signed_url'])
```
Comment (waltjones) [inline src/sourcemaps/command.js:75]: This is a lot of logic of the requester and uploader present in command.js. Ideally, this module should just route the arguments to the correct class to do its work internally. 

Maybe for example, the requester instance can be passed to the `SignedUrlUploader` constructor:
```
const signedUrlUploader = new SignedUrlUploader(requester);
```
Then all of the remaining logic could be encapsulated in something like:
```
signedUrlUploader.upload();
```
There are other reasonable ways, but the idea is to get the logic into the relevant classes.

File: src/sourcemaps/command.js [hunk 6]
```diff
@@ -50,13 +58,32 @@ exports.handler = async function (argv) {
...
-  const uploader = new Uploader({
-    accessToken: argv['access-token'],
-    baseUrl: argv['url-prefix'],
-    codeVersion: argv['code-version']
-  })
+  if (argv['signed-url']) {
+    const requester = new Requester({
+      accessToken: argv['access-token'],
+      baseUrl: argv['url-prefix'],
+      codeVersion: argv['code-version']
```
Comment (waltjones) [inline src/sourcemaps/command.js:65]: If this class will need `argv['dry-run']` why not pass it here.

File: src/common/rollbar-api.js [hunk 7]
```diff
@@ -36,6 +36,20 @@ class RollbarAPI {
...
   }
 
+  async sigendURLsourcemaps(request) {
+
+    const json = JSON.stringify({ version: request.version , prefix_url: request.baseUrl});
+
+    const resp = await this.axios.post(
+      '/signed_url/sourcemaps', json,
+      { headers: {
+        // axios needs some help with headers for form data.
```
Comment (waltjones) [inline src/common/rollbar-api.js:46]: This comment is out of place, as the request here doesn't use form data or its headers.

File: src/sourcemaps/requester.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    const file = 'manifest.json'
+    const outFile = path.join(filePath, file);
+    const data = {
+      projectID: this.projectID,
+      version: this.version,
+      baseUrl: this.baseUrl,
+    };
+    const strData = JSON.stringify(data, null, 2);
+
+    fs.writeFileSync(outFile, strData, (err) => {
```
Comment (waltjones) [inline src/sourcemaps/requester.js:60]: Ideally we wouldn't write to the user's local file system, but if we do, we should probably clean it up when done.

File: src/sourcemaps/signed-url-uploader.js [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+          if (file.validated) {
+            zipFile.addLocalFile(file.mappedFile);
+          }
+        } catch(e) {
+          output.status('Error', e.message)
+        }
+      }
+      try {
+        const outFile = path.join(targetPath, filename);
+        fs.writeFileSync(outFile, zipFile.toBuffer());
```
Comment (waltjones) [inline src/sourcemaps/signed-url-uploader.js:39]: Same idea here as above. We should consider cleaning up temp files.

File: test/common/rollbar-api.test.js [hunk 10]
```diff
@@ -35,6 +35,42 @@ describe('.sourcemaps()', function() {
...
+    };
+
+    const response = await rollbarAPI.sigendURLsourcemaps(request);
+
+    expect(response).to.be.not.null;
+    expect(stub.calledOnce).to.be.true;
+
+    const body = stub.getCall(0).args;
+    expect(body[0]).to.equal('/signed_url/sourcemaps');
+    expect(body[1]).to.be.a('String'); // This is how Chai sees the Buffer type
```
Comment (waltjones) [inline test/common/rollbar-api.test.js:70]: Comment is misplaced here. This isn't a Buffer type.

**Comment by waltjones** [comment]
@pawelsz-rb Core functionality looks good. Two overall things to consider:

It makes sense to write to a file on disk if the zip is too big to keep in memory, but it looks like the zip is fully in memory anyway before writing it. I would consider either looking for a way to do the work fully in memory so you don't need to write temp files only to read them right back in. Or since that memory usage may be unacceptable for real world work loads, look for a way to chunk the work both during compress and during the stream upload. If it's not going to be chunked though, there's probably no upside to writing it to the file system.

Second thing is semicolons. Semicolon use is mixed throughout in this PR. Rollbar js projects in general use semicolons, so that's the style to follow. Maybe we should check the lint settings and see if this check can be enabled.

**Comment by pawelsz-rb** [comment]
> @pawelsz-rb Core functionality looks good. Two overall things to consider:
> 
> It makes sense to write to a file on disk if the zip is too big to keep in memory, but it looks like the zip is fully in memory anyway before writing it. I would consider either looking for a way to do the work fully in memory so you don't need to write temp files only to read them right back in. Or since that memory usage may be unacceptable for real world work loads, look for a way to chunk the work both during compress and during the stream upload. If it's not going to be chunked though, there's probably no upside to writing it to the file system.
> 
> Second thing is semicolons. Semicolon use is mixed throughout in this PR. Rollbar js projects in general use semicolons, so that's the style to follow. Maybe we should check the lint settings and see if this check can be enabled.

thank you for your constructive review, I think it was not that bad as for my very first js project. it's ready for another review

**Comment by waltjones** [comment]
> I think it was not that bad as for my very first js project.

Indeed! Not bad at all. Nice work. :+1:

**Review by waltjones — APPROVED** [review]

**Review by ajtran — APPROVED** [review]

**Pull request merged**